File: library/src/main/java/com/mikepenz/materialdrawer/util/DrawerImageLoader.java
package com.mikepenz.materialdrawer.util;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.widget.ImageView;

/**
 * Created by mikepenz on 24.03.15.
 */
public class DrawerImageLoader {
    public enum Tags {
        PROFILE,
        PROFILE_DRAWER_ITEM,
        ACCOUNT_HEADER
    }

    private static DrawerImageLoader SINGLETON = null;

    private IDrawerImageLoader imageLoader;
    private boolean mHandleAllUris = false;

    private DrawerImageLoader(IDrawerImageLoader loaderImpl) {
        imageLoader = loaderImpl;
    }

    public static DrawerImageLoader init(IDrawerImageLoader loaderImpl) {
        SINGLETON = new DrawerImageLoader(loaderImpl);
        return SINGLETON;
    }

    public static DrawerImageLoader getInstance() {
        if (SINGLETON == null) {
            SINGLETON = new DrawerImageLoader(new AbstractDrawerImageLoader() {
            });
        }
        return SINGLETON;
    }

    public DrawerImageLoader withHandleAllUris(boolean handleAllUris) {
        this.mHandleAllUris = handleAllUris;
        return this;
    }

    /**
     * @param imageView
     * @param uri
     * @param tag
     * @return false if not consumed
     */
    public boolean setImage(ImageView imageView, Uri uri, String tag) {
        //if we do not handle all uris and are not http or https we keep the original behavior
        if (mHandleAllUris || "http".equals(uri.getScheme()) || "https".equals(uri.getScheme())) {
            if (imageLoader != null) {
                Drawable placeHolder = imageLoader.placeholder(imageView.getContext(), tag);
                imageLoader.set(imageView, uri, placeHolder);
            }
            return true;
        }
        return false;
    }

    public void cancelImage(ImageView imageView) {
        if (imageLoader != null) {
            imageLoader.cancel(imageView);
        }
    }

    public IDrawerImageLoader getImageLoader() {
        return imageLoader;
    }

    public void setImageLoader(IDrawerImageLoader imageLoader) {
        this.imageLoader = imageLoader;
    }

    public interface IDrawerImageLoader {
        void set(ImageView imageView, Uri uri, Drawable placeholder);

        void cancel(ImageView imageView);

        Drawable placeholder(Context ctx);

        /**
         * @param ctx
         * @param tag current possible tags: "profile", "profileDrawerItem", "accountHeader"
         * @return
         */
        Drawable placeholder(Context ctx, String tag);
    }
}

File: library/src/main/java/com/mikepenz/materialdrawer/util/DrawerImageLoader.kt
package com.mikepenz.materialdrawer.util

import android.content.Context
import android.graphics.drawable.Drawable
import android.net.Uri
import android.widget.ImageView

/**
 * Created by mikepenz on 24.03.15.
 */
open class DrawerImageLoader private constructor(var imageLoader: IDrawerImageLoader?) {

    private var mHandleAllProtocols = false
    private var mHandledProtocols = listOf<String?>("http", "https")

    enum class Tags {
        PROFILE,
        PROFILE_DRAWER_ITEM,
        ACCOUNT_HEADER
    }

    /**
     * Makes this DrawerImageLoader handle all Uri protocols
     */
    fun withHandleAllProtocols(handleAllProtocols: Boolean): DrawerImageLoader {
        this.mHandleAllProtocols = handleAllProtocols
        return this
    }

    /**
     * @param protocols The Uri protocols which this DrawerImageLoader will handle
     */
    fun withProtocols(vararg protocols: String): DrawerImageLoader {
        this.mHandledProtocols = protocols.toList()
        return this
    }

    /**
     * @param imageView
     * @param uri
     * @param tag
     * @return false if not consumed
     */
    open fun setImage(imageView: ImageView, uri: Uri, tag: String?): Boolean {
        // If we do not handle this protocol we keep the original behavior
        return if (mHandleAllProtocols || uri.scheme in mHandledProtocols) {
            imageLoader?.let {
                val placeHolder = it.placeholder(imageView.context, tag)
                it[imageView, uri, placeHolder] = tag
            }
            true
        } else false
    }

    fun cancelImage(imageView: ImageView) {
        imageLoader?.cancel(imageView)
    }

    interface IDrawerImageLoader {
        @Deprecated("")
        operator fun set(imageView: ImageView, uri: Uri, placeholder: Drawable)

        operator fun set(imageView: ImageView, uri: Uri, placeholder: Drawable, tag: String?)

        fun cancel(imageView: ImageView)

        fun placeholder(ctx: Context): Drawable

        /**
         * @param ctx
         * @param tag current possible tags: "profile", "profileDrawerItem", "accountHeader"
         * @return
         */
        fun placeholder(ctx: Context, tag: String?): Drawable
    }

    companion object {

        private var SINGLETON: DrawerImageLoader? = null

        fun init(loaderImpl: IDrawerImageLoader): DrawerImageLoader {
            SINGLETON = DrawerImageLoader(loaderImpl)
            return SINGLETON as DrawerImageLoader
        }

        val instance: DrawerImageLoader
            get() {
                if (SINGLETON == null) {
                    SINGLETON = DrawerImageLoader(object : AbstractDrawerImageLoader() {

                    })
                }
                return SINGLETON as DrawerImageLoader
            }
    }
}
